Add a spec for search, #605 (#3299)
microsoft/terminal: doc/specs/#605 - Search/spec.md

## Solution Design

Our ultimate goal is to provide both search within one tab and search from all tabs experiences. But we can start with one-tab search. The search experience should have following features:

1. The search is triggered by KeyBindings. A new setting property named "Search" will be enabled in the Json file. The user can set their own key bindings for search. The default is "ctrl + F". 
2. The user search in a XAML AutoSuggestBox on the top right corner of the Terminal window. 
3. We can have multiple search methods. The simplest one is text exact match. Other match methods include case-sensitive exact match and regex match. In the first phrase, we will focus on case sensitive/insensitive text exact match. 
4. If currently there is no active selection, the search starts from the line that the cursor is on. If there is an active selection, we start from the previous or the next text of the selected text. We automatically go around if we reach the start point of the search. 

> 5. The search dialog should not block terminal's view. 

Author: Added

1. Search is performed on a XAML AutoSuggestBox. Once the user click the "Find" icon or press Enter, we start from the cursor (the current input line on the Terminal) or the current selection, and try to find the exact text in the text buffer. The nearest searched one will be selected. Then the search start point will be set to the selected text. The next time "Find" button is clicked, the search will start before or after the previous searched text.
2. The user can choose to search up or down by choosing up arrow or down arrow buttons. A blue border will appear on the chosen button. If the user does not click the arrows buttons, the default direction is up. 
3. The user can choose to do case sensitive or insensitive match by checking a check box. 
4. If the user click the "X" button or press Esc, the search stopped and the search box disappears. In phrase one we do not store any state. 
5. Once the search box is closed (exiting search mode), the selection will still be there. This coincides with the current VS Code and cmd experience. To get rid of the selection, the user can just click other area of the window. 

## UI/UX Design

![SearchBox mockup](images/SearchBoxControl.png)

> Above is the SearchBoxControl. The two buttons with up/down arrows controls the search direction, a blue border will appear on the current selected direction. The checkbox, if checked, means that we are searching case-sensitivily. 

Reviewer: Looks like a comment I had before got lost:

>  A blue border will appear on the chosen button. 

I'm no PM, but IMO I'd avoid being _too_ specific with calling out the UI design here. Instead, try "The chosen button will be styled to indicate it is focused". That way the spec can remain prescriptive even as UI paradigms change.

@cinnamon-msft can call me out though if that's not right

Author: clicked.

### Accessibility

This feature could help multitaksing terminal users. 

### Security

This feature should not introduce any new security issues.

### Reliability

This feature enable users to search for text in the terminal input/output history. This is a widely-used feature in most editors and thus improve the reliability of Terminal once added. 

### Compatibility

This feature won't break existing features of Terminal.

### Performance, Power, and Efficiency

This feature only launches in need. It does not impact the performance of Terminal. 

## Potential Issues

1. The search bar should not block command line view. 
2. Search should not block any program the Terminal is currently executing.

## Future considerations

In version 1, we want realize a case sensitive/insensitive exact text match. But we may consider the following features in version 2:

1. Add "Find" button in dropdown menu to trigger search. 
2. Search from all tabs. For Version 1 we just want to realize search within one tab. However, the community also requests search from all tabs. We put in our goals for Version 2. 
3. Regular experssion match. 
4. Search history.

> 5. High-light while you type. 

Reviewer: Another idea for future consideration: Emphasizing all the other matches in the buffer with an outline or something, similar to how sublime works:
![image](https://user-images.githubusercontent.com/18356694/69557540-6266a480-0f6c-11ea-995e-c8a34234b8a7.png)

Though, this probably needs some cooperation with the renderer to support rendering like this.

Author: Added in the Future considerations.

---
author: Kaiyu Wang KaiyuWang16/kawa@microsoft.com
created on: 2019-10-16
last updated: 2019-10-16
issue id: #605
---

# Search in Terminal

## Abstract

This spec is for feature request #605 "Search". It goes over the details of a new feature that allows users to search text in Terminal, within one tab or from all tabs. Expected behavior and design of this feature is included. Besides, future possible follow-up works are also addressed. 

## Inspiration

One of the superior features of iTerm2 is it's content search. The search comes in two variants: search from active tab and search from all tabs. In almost any editor, there is an roughly equivalent string search. We also want to realize search experience in Terminal. There will be two variants, search within one tab or from multiple tabs. We will start with one-tab search implementation. 

## Solution Design

Our ultimate goal is to provide both search within one tab and search from all tabs experiences. But we can start with one-tab search. The search experience should have following features:

> 1. The search is triggered by KeyBindings. A new setting property named "openFind" will be enabled in the Json file. The user can set their own key bindings for search. The default is "ctrl + F". 

Reviewer: ```suggestion
1. The search is triggered by KeyBindings. A new setting property named "openFind" will be enabled in the Json file. The user can set their own key bindings for search. The default is <kbd>ctrl+f</kbd>. 
```

Also, I'm a little worried that the default is <kbd>ctrl+f</kbd>. We've been generally trying to standardize on <kbd>ctrl+shift+</kbd> as the modifiers for most default keybindings.

Author: I will change to Ctrl+Shift+F

2. The user search in a XAML TextBox, which is contained in a custom SearchBoxControl. The default position of the search box is the top right corner, the user can also move it to bottom right corner to avoid blocking text. 
3. We can have multiple search methods. The simplest one is exact text match. Other match methods include case-sensitive exact match and regex match. In the first phrase, we will focus on case sensitive/insensitive text exact match. 

> 4. If currently there is no active selection, the search starts from the last line of the viewport. If there is an active selection, we start from the previous or the next text of the selected text. We automatically go around if we reach the start point of the search. 

Reviewer: I'm a little confused here:
> If currently there is no active selection, the search starts from the last line of the viewport

Is this the last line of the visible viewport, or the last line of the "mutable" viewport (i.e. the last line of the buffer with text in it? Or does the search start from the cursor's position?

Author: I use  the last line of the mutableViewport

Reviewer: Let's make sure to call that out specifically in the spec then :)

5. The search dialog should not block terminal's view. In phrase one, this is achieved by make search box movable. The user can move the search box to the bottom right corner.
6. The user should be able to fully interact with the terminal when the search box is on screen. 
7. For accessibility concerns, the user should be able to navigate all the iteractive elements on the search box using keyboard tab if the search box is focused. Searchbox could be created and closed with keyboard bindings. Close is usually binding to Esc. 

> 8. To avoid searchBox from blocking the terminal window, the search box could be moved to the bottom right by clicking a button. 

Reviewer: 8 seems like it should really be a part of 5 here.

Author: removed 8.

> 8. To avoid searchBox from blocking the terminal window, the search box could be moved to the bottom right by clicking a button. 

Author: Yes, added.

> Conhost already has a module for search. It realizes case sensitive or insensitive exact text match search, and it provides methods to select the found word. However, we want to make search as a shared component between Terminal and Console host. Now search module is part of Conhost, and its dependencies include BufferOut and some other types in ConHost such as SCREEN_INFORMATION. In order to make Search a shared component, we need to remove its dependency on ConHost types. BufferOut is already a shared component, but we need to make sure there is no other Conhost dependency.

Reviewer: Protip: you can use backticks ``` ` ``` to surround text and format it `like code`. This can be especially helpful for emphasizing that a word is from code, rather than just a normal word.

```suggestion
Conhost already has a module for search. It implements case sensitive or insensitive exact text match search, and it provides methods to select the found word. Currently, search is part of Conhost, and its dependencies include `BufferOut` and some other types in ConHost such as `SCREEN_INFORMATION`.

However, we want to make search as a shared component between the Terminal and Conhost. In order to make Search a shared component, we will need to remove its dependency on ConHost-specific types. `BufferOut` is already a shared component, but we need to make sure there is no other `host` dependencies.
```

I'd suggest applying these in other places in this spec, especially on things like `TermControl`, `SearchBoxControl`, `UserControl`, and other XAML types

Author: Added

We will create a SearchBoxControl Xaml UserControl element. When a search process begins, a SearchBoxControl object will be created and attach to TermControl root grid. In other words, one SearchBox is added for each TermControl. The reasons for this design is:

> 1. Each TermControl object is a Terminal Window and has a individual text buffer. In phrase 1 we are going to search witin the current terminal text buffer. 

Reviewer: From a lost comment:
> One of the big concerns I still have with this spec is regarding per-terminal search vs global search.
> 
> > One of the superior features of iTerm2 is it's content search. The search comes in two variants: search from active tab and **search from all tabs**
> 
> (emph. added)
> 
> With the current design of a per-terminal find dialog, how do we transition to a global (search all terminals) model? I know that global search is out of scope for this first iteration, but if the find dialog is inherently tied to the terminal instance it's searching, how do we then re-use this dialog for searching many terminal instances?
>

Author: I discussed with Dustin about in which directory should implement the SearchBoxControl, and we decided to add this control in TerminalControl/ dir. In this way, developers could use this UserControl as an individual component. I think if we wants to implement multi-tab search, we can attach a SearchBoxControl object to the App or TerminalPage, we can provide the text buffer of the current focused terminal. But we need to change the search algorithm.

Reviewer: Could we add this to the "Future Considerations" section on search from all tabs? I think that would help answer questions. Just add something about how we could later add a modified `SearchBoxControl` to `TerminalPage` directly to enable a "search from all tabs" option in the future

Author: Added.

> 1. Each TermControl corresponds to an iterative Terminal Window and has a individual text buffer. In phrase 1 we are going to search witin the current terminal text buffer. 

Reviewer: > **iterative** Terminal Window

What does that mean?

Author: Changed to "Each TermControl object is a Terminal Window....." to make it clearer.

2. If we put the search box under TerminalApp, then the search can only happen on the current focused Terminal. 
3. If the community does not like the current design, we can lift SearchBox to a higher level. 

### Search process
1. The user press Ctrl+F (or user's custom key binding) to open the search box. Focus will move to the TextBox. 
2. Search is performed on a XAML TextBox. Once the user presses Enter, we start to search from the last line of the current viewport or the current selection, and try to find the exact text in the text buffer. The nearest searched one will be selected. Then the search start point will be set to the selected text. The next time "Find" button is clicked, the search will start before or after the previous searched text.

> 3. The user can choose to search up or down by choosing up arrow or down arrow buttons. The chosen button will be styled to indicate it is clicked. If the user does not click the arrows buttons, the default direction is up. 

Reviewer: ```suggestion
3. The user can choose to search up or down by choosing up arrow or down arrow buttons. The chosen button will be styled to indicate it is selected. If the user does not click the arrows buttons, the default direction is up. 
```

> 3. The user can choose to search up or down by choosing up arrow or down arrow buttons. The chosen button will be styled to indicate it is clicked. If the user does not click the arrows buttons, the default direction is up. 

Reviewer: If the user clicks on a arrow button, does it simply change the direction of the search, or does it also execute a search (moving the selection to the next matched text)?

Author: For now, it only simply change the direction of the search. But and search execution could be added. I will include it in my PR update.

4. The user can choose to do case sensitive or insensitive match by checking a check box. 
5. If the search box is focused, the user can navigte all the elements on the search box using tab. When selected, press Enter equals to click. 

> 6. If the user click the "X" button or press Esc, the search stopped and the search box disappears and focus will move back to Terminal. In phrase one we do not store any state. 

Reviewer: ```suggestion
6. If the user click the "X" button or press <kbd>Esc</kbd>, the search stopped and the search box disappears and focus will move back to Terminal. In phrase one we do not store any state. 
```

7. Once the search box is closed (exiting search mode), the selection will still be there. This coincides with the current VS Code and cmd experience. To get rid of the selection, the user can just click other area of the window.

> 

Reviewer: What happens if the user clicks on the terminal content itself, drawing focus from the find dialog? Does the dialog stay open? Does it dismiss the current selection, but keep the dialog open? Can the user scroll the terminal content, or otherwise input text into the terminal while the dialog is open?

Similarly, what happens if the user switches tabs while the find dialog is open? Does re-focusing the tab move focus to the find dialog of the focused `TermControl`, or does it move focus to the terminal itself?

Author: 1. What happens if the user clicks on the terminal content itself, drawing focus from the find dialog? Does the dialog stay open? Does it dismiss the current selection, but keep the dialog open? 

Yes, the search dialog stays open. The current selection is dismissed because of the click. 

2. Can the user scroll the terminal content, or otherwise input text into the terminal while the dialog is open?

Yes, the user can scroll or input text into the terminal content even if the search dialog is open. The user can have full interaction with the terminal. 

3. what happens if the user switches tabs while the find dialog is open? Does re-focusing the tab move focus to the find dialog of the focused TermControl, or does it move focus to the terminal itself?

Neither. It will not focus on the TermControl. I will fix this.

Author: Update: I synced and with the new changes, switching between tabs is working very well. Once a tab is re-focused with the search dialog open, focus will still be moved to Terminal.

Author: Added.

> 7. Once the search box is closed (exiting search mode), the selection will still be there. This coincides with the current VS Code and cmd experience. To get rid of the selection, the user can just click other area of the window.

Reviewer: Lots of this seems like it fits more in the "UI/UX Design" (user interface/user experience) part of the spec. Especially points 3, 4, 5, 6. For those, I'd try to remove references to the UI itself here, and just be more descriptive about what happens algorithmic-ly behind the scenes here.

Author: Could I move the search process part to UI/UX design ?

## UI/UX Design

> ![SearchBox mockup](images/SearchBoxControl.png)

Author: Yes, we have RotateTransform: https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Media.RotateTransform

> Above is the SearchBoxControl. The two buttons with up/down arrows controls the search direction, it will be styled once clicked. The checkbox, if checked, means that we are searching case-sensitivily. The Dockbottom button on the right of the checkbox is for moving the search box to top/bottom. 

Author: changed

> ![SearchBox mockup, arrow button clicked](images/SearchBoxUpSelected.png)

Reviewer: From another lost comment:

> 
> the UI seems a little tall - one row for each of the close button, the search box, and the case sensitivity chckbox means that this UI is now obscuring 3 rows of terminal content. Perhaps we could collapse it to a single row more like VsCode?
>        ![image](https://user-images.githubusercontent.com/18356694/68127057-b553bb80-feda-11e9-99e5-6e64dea9c509.png)
> 
>  I would consider calling out these UI as _examples_ of what the UI could look like, but not necessarily what it will _definitely_ look like. Then we're less tied to content of the spec in whatever we end up implementing 😋

Author: I think if we put everything on one line, the search box would be very wide and it may block text more frequently ?

Reviewer: Ooh, that's a good point. Though, the only real problem we'll have for it blocking text is for matches that are on the very first line of the buffer, right? Considering if the match is on any other line, the user can just scroll up to un-block text. 

Regardless, whatever the design is that's chosen, we should probably include arguments _for_ and _against_ each style in the spec, with a reason given for picking one style over the other.

Author: Added, and I think if the community provides feedbacks in the future that single-row layout is prefered, we can change  in the future.

> The search box defaults to be on the top right corner of the Terminal window. If the current tab is split into panes, each pane will have a individual searchbox. The user can move it to bottom to avoid it blocking too much screen contents. 

Author: changed

> 

Author: No, I wonder if we want this in Phrase 1 ?

> 

Reviewer 2: I played around with the implementation branch a bit and it feels really weird without it. Honestly, I think it should be in Phase 1.

> 

Author: Added

4. Found text will be selected. 
5. The search will start from the active selected text (inclusive) if there is one, or where the cursor is. 
5. The search will automatically go around when it reaches the starting point.
7. The search box can be positioned at either the top left or the bottom left of the terminal. 
8. The user can use Tab to navigate all the elements in the search box. 

### Accessibility

The user should be able to use search by keyboard only.
Once the searchbox is focused, the user can navigate between elements in the search box using Tab. And "click" using Enter. 

### Security

This feature should not introduce any new security issues.

### Reliability

1. The key input of Terminal command line and the search box should be separated. search box should not block interaction with the command line when it is open. 
2. The search box should not block text. This issue could be addressed by make search box movable. 

### Compatibility

This feature won't break existing features of Terminal.

### Performance, Power, and Efficiency

This feature only launches in need. It does not impact the performance of Terminal. 

## Potential Issues

> 1. The search bar should not block command line view. 

Author: I do not have any potential issues for now, but we can add at any time.

2. Search should not block any program the Terminal is currently executing.

## Future considerations

In version 1, we want realize a case sensitive/insensitive exact text match. But we may consider the following features in version 2:

1. Add "Find" button in dropdown menu to trigger search. 
2. Search from all tabs. For Version 1 we just want to realize search within one tab. However, the community also requests search from all tabs. We put in our goals for Version 2. 
3. Regular experssion match. 
4. Search history.

> 5. High-light while you type. Emphasizing all the other matches in the buffer with an outline or selection with another color. 

Reviewer: Add some more context here. Why would we want this feature? Why is it not being considered part of this spec currently? (this comment really applies to each of the bullet points in this list)

Author: Added

---
author: Kaiyu Wang KaiyuWang16/kawa@microsoft.com
created on: 2019-10-16

> last updated: 2019-10-16

Reviewer 3: ```suggestion
last updated: 2019-12-10
```

## Inspiration

One of the superior features of iTerm2 is it's content search. The search comes in two variants: search from active tab and search from all tabs. In almost any editor, there is an roughly equivalent string search. We also want to realize search experience in Terminal. There will be two variants, search within one tab or from multiple tabs. We will start with one-tab search implementation. 

## Solution Design

Our ultimate goal is to provide both search within one tab and search from all tabs experiences. But we can start with one-tab search. The search experience should have following features:

1. The search is triggered by KeyBindings. A new setting property named "find" will be enabled in the Json file. The user can set their own key bindings for search. The default is <kbd>ctrl+shift+f</kbd>. 
2. The user search in a XAML TextBox, which is contained in a custom `SearchBoxControl`. The default position of the search box is the top right corner.

> 3. We can have multiple search methods. The simplest one is exact text match. Other match methods include case-sensitive exact match and regex match. In the first phrase, we will focus on case sensitive/insensitive text exact match. 

Reviewer 3: ```suggestion
3. We can have multiple search methods. The simplest one is exact text match. Other match methods include case-sensitive exact match and regex match. In the first phase, we will focus on case sensitive/insensitive text exact match. 
```

phase, not phrase

3. The user can choose to search up or down. Search module realizes this, we just need to set a boolean flag. Default is search up. 
4. The user can choose to do case sensitive or insensitive match. This also realized by Search module by setting a boolean flag. Default is search case-insensitively. 
5. Tab navigation is realized by XAML. We just need to set TabNavigation="Cycle" in `SearchBoxControl`. 
6. If the user click the "X" button or press <kbd>Esc</kbd>, the search box will disappear and the object will be destructed and detached from the `TermControl` XAML tree. In phrase one we do not store any state. 
7. We need to guarantee full interaction with the terminal when the search box is open. To achieve this, search box and terminal input should be separated. If the current keyboard focus is on the search box, then keydown events will be handled on "search box level". 

## UI/UX Design

> ![SearchBox mockup](images/SearchBoxControl.png)

Reviewer 3: It seems like your screenshot could be cropped a lot closer...

Above is the `SearchBoxControl`.
  - The two buttons with up/down arrows controls the search direction, Each button will be styled to indicate which search direction is currently selected. 
  - The button with a "Aa" icon, if pressed, means that we are searching case-sensitivily. 
  - The current style puts all elements - the `X` button, the text box and the search pattern control buttons on one single line. This ensures that the `SearchBoxControl` won't be too high and block terminal text. This is similar with VSCode. Another possible layout style is to put elements in multiple layers. This will occupy more lines, but the search dialog will narrower. Considering that there is not many elements, we do not need multiple layers. 

![SearchBox mockup, arrow button clicked](images/SearchBoxUpSelected.png)

> The search box defaults to be on the top right corner of the Terminal window. If the current tab is split into panes, each pane will have a individual searchbox. The user can quick toggle the search box between the top and bottom to avoid it blocking too much screen contents. 

Reviewer 3: The part about toggling the location is no longer correct.

Author: removed.

#### Search process
1. The user presses <kbd>ctrl+shift+f</kbd> (or user's custom key binding) to open the search box. Focus will move to the TextBox. 
2. Search is performed on a XAML TextBox. Once the user presses Enter or click up/down arrow button, the search starts and searched text will be selected. Next search will be performed beginning from the current selection and go towards up/down. 
3. The user can choose to search up or down by selecting up arrow or down arrow buttons. The chosen button will be styled to indicate it is selected. If the user does not click the arrows buttons, the default direction is up. 
4. The user can choose to do case sensitive or insensitive match by checking a check box. The default is case insensitive. 

> 5. If the search box is focused, the user can navigte all the elements on the search box using tab. When selected, press Enter equals to click. 

Reviewer 3: ```suggestion
5. If the search box is focused, the user can navigate all the elements on the search box using tab. When selected, press Enter equals to click. 
```

navigte -> navigate

1. Add "Find" button in dropdown menu to trigger search. This enables the search feature to be operated with mouse only. However, this is not required by Accessibility so we do not cover this in phrase one. 
2. Search from all tabs. For Version 1 we just want to realize search within one tab. However, the community also requests search from all tabs. This may require a big change to the search algorithm, but it is not seen as a popular use scenario, so we put it future phrase. To implement multi-tab search, we can let TerminalPage or App own a `SearchBoxControl` object, and provide the text buffer of the current focused terminal. We need to change the search algorithm.  
3. Regular expression match. This is a useful search pattern and is implemented in some editors. However, this use scenario is not used as much as exact text search, thus, we put it in future phrase.
4. Search history. Sometimes users would do the same search for several times, thus, storing the search history is useful. This is not realized by VSCode so it would be a good highlighting point in the future. 

> 5. High-light while you type. Emphasizing all the other matches in the buffer with an outline or selection with another color. This provides a clearer view of searched text. But we need to change the search and selection algorithm, so we put it in the future phrase. 

Reviewer 3: Let's add consideration number 6. Some text editors let the user resize the search box, like this:

![image](https://user-images.githubusercontent.com/14316954/70576240-46165a80-1b5d-11ea-882d-6b01dad97097.png)

the dark gray handle on the left is a size handle.

We can mention that if the community desires it we may add a similar feature?

---
author: Kaiyu Wang KaiyuWang16/kawa@microsoft.com
created on: 2019-10-16
last updated: 2019-10-16
issue id: #605
---

# Search in Terminal

## Abstract

This spec is for feature request #605 "Search". It goes over the details of a new feature that allows users to search text in Terminal, within one tab or from all tabs. Expected behavior and design of this feature is included. Besides, future possible follow-up works are also addressed. 

## Inspiration

One of the superior features of iTerm2 is it's content search. The search comes in two variants: search from active tab and search from all tabs. In almost any editor, there is an roughly equivalent string search. We also want to realize search experience in Terminal. There will be two variants, search within one tab or from multiple tabs. We will start with one-tab search implementation. 

## Solution Design

Our ultimate goal is to provide both search within one tab and search from all tabs experiences. But we can start with one-tab search. The search experience should have following features:

> 1. The search is triggered by KeyBindings. A new setting property named "openFind" will be enabled in the Json file. The user can set their own key bindings for search. The default is <kbd>ctrl+shift+f</kbd>. 

Author: Changed to "Find"

> 2. The user search in a XAML TextBox, which is contained in a custom `SearchBoxControl`. The default position of the search box is the top right corner, the user can also move it to bottom right corner to avoid blocking text. 

Reviewer 4: Other text editors don't move the search box around. If text is blocked by the search box, nothing happens and it stays blocked (the user can scroll or just close the search box). Therefore, I don't think we need to change anything for this case and we can leave it in the top right.

Author: removed this feature

3. We can have multiple search methods. The simplest one is exact text match. Other match methods include case-sensitive exact match and regex match. In the first phrase, we will focus on case sensitive/insensitive text exact match. 

> 4. If currently there is no active selection, the search starts from the last line of the mutableViewport. If there is an active selection, we start from the previous or the next text of the selected text. We automatically go around if we reach the start point of the search. 

Reviewer 4: The only additional indicator I can think of is having a count such as "1 of 24". Otherwise, the indicator would just be the scrollbar.

> 4. If currently there is no active selection, the search starts from the last line of the mutableViewport. If there is an active selection, we start from the previous or the next text of the selected text. We automatically go around if we reach the start point of the search. 

Author: I think this may be a good point for phrase 2, together with "Select as you type".

> 5. The search dialog should not block terminal's view. In phrase one, this is achieved by make search box movable. The search box could be moved to the bottom right by clicking a button. It will be moved back to the top right corner by clicking the button again. The button contains a `DockBottom` icon and it will turn upside-down when the search box is on the bottom. 

Reviewer 4: I would also rather this feature be cut out.

> 6. The user should be able to fully interact with the terminal when the search box is on screen. 

Author: Discussed offline and nobody else does it.

![SearchBox mockup, arrow button clicked](images/SearchBoxUpSelected.png)

The search box defaults to be on the top right corner of the Terminal window. If the current tab is split into panes, each pane will have a individual searchbox. The user can quick toggle the search box between the top and bottom to avoid it blocking too much screen contents. 

#### Search process
1. The user presses <kbd>ctrl+shift+f</kbd> (or user's custom key binding) to open the search box. Focus will move to the TextBox. 
2. Search is performed on a XAML TextBox. Once the user presses Enter or click up/down arrow button, the search starts and searched text will be selected. Next search will be performed beginning from the current selection and go towards up/down. 
3. The user can choose to search up or down by selecting up arrow or down arrow buttons. The chosen button will be styled to indicate it is selected. If the user does not click the arrows buttons, the default direction is up. 
4. The user can choose to do case sensitive or insensitive match by checking a check box. The default is case insensitive. 

> 5. If the search box is focused, the user can navigte all the elements on the search box using tab. When selected, press Enter equals to click. 

Reviewer 2: ```suggestion
5. If the search box is focused, the user can navigate all the elements on the search box using tab. When selected, pressing Enter or Space equals to click. 
```
I think it's standard that the Spacebar invokes the XAML Control

6. If the user click the "X" button or press <kbd>Esc</kbd>, the search stopped and the search box disappears and focus will move back to Terminal.
7. Once the search box is closed (exiting search mode), the selection will still be there. This coincides with the current VS Code and cmd experience. To get rid of the selection, the user can just click other area of the window.
8. If the user clicks on the terminal when the search box is open, it will draw focus back to the terminal from the search box. The search box will still stay open. 

> 9. The user can interact with the terminal when the search box is open, which means that ther user can scroll the terminal content, or input text. 

Reviewer 2: ```suggestion
9. The user can interact with the terminal when the search box is open, which means that ther user can scroll the terminal content, or input text when the focus is on the terminal control. 
```

> 10. If the user switches tabs while the search box is open, the re-focusing will move focus to the terminal. 

Reviewer 2: what do you mean by re-focusing? This point isn't clear.

## Capabilities

1. The user can search exact matched text in the text buffer of the Terminal Screen. 
2. The user can choose to search case sensitively and insensitively. 
3. The user can search up or down. 
4. Found text will be selected. 
5. The search will start from the active selected text (inclusive) if there is one, or where the cursor is. 
5. The search will automatically go around when it reaches the starting point.
7. The search box can be positioned at either the top left or the bottom left of the terminal. 
8. The user can use Tab to navigate all the elements in the search box. 

### Accessibility

The user should be able to use search by keyboard only.
Once the searchbox is focused, the user can navigate between elements in the search box using Tab. And "click" using Enter. 

### Security

This feature should not introduce any new security issues.

### Reliability

> 1. The key input of Terminal command line and the search box should be separated. search box should not block interaction with the command line when it is open. 

Reviewer 2: ```suggestion
1. The key input of Terminal command line and the search box should be separated. Search box should not block interaction with the command line when it is open. 
```

> 2. The search box should not block text. This issue could be addressed by make search box movable. 

Reviewer 2: I think it should _minimally_ block the terminal control. To my earlier comment, if we want to make the search box movable, we should be able to move it _outside_ of the Terminal window. Though that may not be a viable option today, I think we just have to hold off until then.

### Compatibility

This feature won't break existing features of Terminal.

### Performance, Power, and Efficiency

This feature only launches in need. It does not impact the performance of Terminal. 

## Potential Issues
                                                              

> 1. If the terminal window is not wide enough for the search box to be visible, the buttons on the right of the `TextBox` will become invisible, but the `TextBox` is still visible and the window could not be narrower than the `TextBox`. This is similar to the behavior of other editors. 

Reviewer 2: It would be nice to include an image here of what this would look like. 

@cinnamon-msft can help with that.

2. If the terminal window is not high enough for the search box to be visible, the whole terminal screen, inlcuding the `SearchBoxControl` can disappear. This is similar to the behavior of other editors.

> ## Future considerations

Reviewer 2: Can you create follow-up work items for these and link them here?

In version 1, we want realize a case sensitive/insensitive exact text match. But we may consider the following features in version 2:

> 1. Add "Find" button in dropdown menu to trigger search. This enables the search feature to be operated with mouse only. However, this is not required by Accessibility so we do not cover this in phrase one. 

Reviewer 2: ```suggestion
1. Add "Find" button in dropdown menu to trigger search.
```
The dropdown menu is accessible via the keyboard. So the accessibility concern doesn't exist.

Author: Actually, I mean "the user can use mouse to launch the search dialog" here.

2. Search from all tabs. For Version 1 we just want to realize search within one tab. However, the community also requests search from all tabs. This may require a big change to the search algorithm, but it is not seen as a popular use scenario, so we put it future phrase. To implement multi-tab search, we can let TerminalPage or App own a `SearchBoxControl` object, and provide the text buffer of the current focused terminal. We need to change the search algorithm.  
3. Regular expression match. This is a useful search pattern and is implemented in some editors. However, this use scenario is not used as much as exact text search, thus, we put it in future phrase.
4. Search history. Sometimes users would do the same search for several times, thus, storing the search history is useful. This is not realized by VSCode so it would be a good highlighting point in the future. 
5. High-light while you type. Emphasizing all the other matches in the buffer with an outline or selection with another color. This provides a clearer view of searched text. But we need to change the search and selection algorithm, so we put it in the future phrase. 

> 6. Enable <kbd>shift</kbd> + <kbd>Enter</kbd> for search in the opposite direction. This helps users to search in either direction with simple key bindings. However, tab navigation can help users to change search direction without mouse, so we put it in the future phrase. 

Reviewer 2: Again, this should be a Phase 1 feature.

Author: added

---
author: Kaiyu Wang KaiyuWang16/kawa@microsoft.com
created on: 2019-10-16
last updated: 2019-10-16
issue id: #605
---

# Search in Terminal

## Abstract

This spec is for feature request #605 "Search". It goes over the details of a new feature that allows users to search text in Terminal, within one tab or from all tabs. Expected behavior and design of this feature is included. Besides, future possible follow-up works are also addressed. 

## Inspiration

One of the superior features of iTerm2 is it's content search. The search comes in two variants: search from active tab and search from all tabs. In almost any editor, there is an roughly equivalent string search. We also want to realize search experience in Terminal. There will be two variants, search within one tab or from multiple tabs. We will start with one-tab search implementation. 

## Solution Design

Our ultimate goal is to provide both search within one tab and search from all tabs experiences. But we can start with one-tab search. The search experience should have following features:

> 1. The search is triggered by keybindings "Ctrl + F". This coincides with other editors. In the future we will also consider adding a "Find" in the dropdown menu. 

Reviewer: I'd probably avoid calling out a specific keybinding here. Instead, I'd say that we'll add a `ShortcutAction` that can be used to toggle the Find dialog, bound to <kbd>ctrl+F</kbd> by default.

I'd also refrain from mentioning the "find in the dropdown menu" feature here - You discuss it in greater detail in the Further Considerations section, let's just contain the discussion to that section.

Author: changed

> 2. The user search in a XAML AutoSuggestBox on the top right corner of the Terminal window. 

Reviewer: Why an `AutoSuggestBox`? Why not just a `TextBox`?

Author: I discussed with Kayla, and she would like the search box to be a AutoSuggest box because in the future we may consider auto suggest functionality.

Reviewer: When we do have autocomplete suggestions, how will those appear with the "case sensitivity" checkbox beneath the search box? Will suggestions appear on top of the checkbox, partially obscuring it?

3. We can have multiple search methods. The simplest one is text exact match. Other match methods include case-sensitive exact match and regex match. In the first phrase, we will focus on case sensitive/insensitive text exact match. 
4. If currently there is no active selection, the search starts from the line that the cursor is on. If there is an active selection, we start from the previous or the next text of the selected text. We automatically go around if we reach the start point of the search. 
5. The search dialog should not block terminal's view. 

Conhost already has a module for search. It realizes case sensitive or insensitive exact text match search, and it provides methods to select the found word. However, we want to make search as a shared component between Terminal and Console host. Now search module is part of Conhost, and its dependencies include BufferOut and some other types in ConHost such as SCREEN_INFORMATION. In order to make Search a shared component, we need to remove its dependency on ConHost types. BufferOut is already a shared component, but we need to make sure there is no other Conhost dependency.

> Search process:

Reviewer: ```suggestion
### Search process
```

> Search is performed on a XAML AutoSuggestBox. Once the user click the "Find" icon, we start from the cursor (or the current selection), and try to find the exact text in the text buffer. The nearest searched one will be selected. And we set the search start point to the selected text. The next time "Find" button is clicked, the search will start before or after the previous searched text.

Author: Sorry for misunderstanding. The starting point may be the cursor or the current selection, and the user can choose to go up or down, I will emphasize this in the spec.

> The user can choose to search up or down by choosing up arrow or down arrow buttons.

Reviewer: > up arrow or down arrow buttons
What if the user just hits <kbd>Enter<kbd> while the search dialog is open? What direction does it search in?

Author: Thanks for this reminding. I think we should set a default direction (Up).

> The user can choose to do case sensitive or insensitive match by clicking a button.

Reviewer 2: What's the user story for NOT using the mouse? How can you make this functionality available by ONLY using a keyboard? Do we want keybindings for these actions? Are they customizable? This may be an entirely separate section but definitely a must.

Author: Tab navigation added. Close is bound to Esc.

> If the user click the "X" button, the search stopped and the search box disappears. In phrase one we do not store any state, but in the future we can consider storing the search history. 

Reviewer 2: Does a selection persist when you exit find mode?

I would also move the statement about "in the future" to the later section on "future considerations".

Author: In VS code, the seleciton will still be there if the search box is closed. We can also follow this pattern.

> If the user click the "X" button, the search stopped and the search box disappears. In phrase one we do not store any state, but in the future we can consider storing the search history. 

Reviewer: > In phrase one we do not store any state, but in the future we can consider storing the search history.  

I'd probably remove that from this section, and leave it for discussion in the "Further Considerations" section.

Author: Removed

> ## UI/UX Design

Reviewer: What if the user wants to dismiss the dialog without the mouse? Can they use the same keybinding to hid the dialog again? What if they want to use <kbd>Esc</kbd> to dismiss the dialog?

Author: In VS code and cmd, Esc is binding to the close action, I think we can also follow this pattern.

Reviewer: We should make sure to add that to the spec then :)

Author: Added.

> ![Sol Design](images/AutoSuggestBox.png)

Reviewer: ```suggestion
![Search Box Mockup](images/AutoSuggestBox.png)
```

> Above is the XAML AutoSuggestBox. We also need to add at least two more buttons, one for case sensitive/insensitive switch, and one for search direction switch. We can consider adding the buttons on the right side of the AutoSuggestBox.

Reviewer: If we're going to include a mockup at all, lets include a mockup with the completed design if at all possible. I'd recommend [XAML Studio](https://www.microsoft.com/en-us/p/xaml-studio/9ntls214tkmq?activetab=pivot:overviewtab) for quick XAML mockups.

> The search box should be on the top right corner of the Terminal window. We need to avoid it blocking too much screen contents. 

Reviewer 2: @cinnamon-msft Do we want the box to be drawn on top of the terminal or to push it down (and not cover anything)?

@KaiyuWang16 Can the user still interact with the terminal when the box is on the screen? Here's a few ideas:
- Full interaction: user can interact with the terminal as if the box was not present
- Limited interaction: user can still scroll the terminal, but cannot type commands or keybindings if the box is present
- No interaction: user cannot interact with the terminal control when the find box is present.
I'm tempted to say that "full interaction" is best but I want to hear people's thoughts.

Author: I also think full interaction is preferred.

Reviewer: I think anything other than "full-interaction" is a non-starter here

Author: Our current design is to put the search box on the top right corner of the TermControl screen. This coincides with VS code.

> The search box should be on the top right corner of the Terminal window. We need to avoid it blocking too much screen contents. 

Reviewer: > We need to avoid it blocking too much screen contents

What happens when the string that's found is immediately under the find box, so it's effectively hidden?

Author: As discussed, this could be avoided by moving the search box to the bottom of the terminal. A new DockBottom button is added.

> The search box should be on the top right corner of the Terminal window. We need to avoid it blocking too much screen contents. 

Reviewer: > The search box should be on the top right corner of the Terminal window

Of the window, or of a Pane? What happens when the user has multiple panes open, and wants to find in one of them? Should there be one global find box for the application, or one per TermControl? If it's global, should the user be able to search for text in one pane, then switch to a new pane, (while the dialog is open) and execute the same search again?

(and similar questions with switching between tabs)

Author: Sorry for misunderstanding. My current design is that each TermControl owns a searchbox. Thus, each pane has a search box.

## Capabilities

1. The user can search exact matched text in the text buffer of the Terminal Screen. 
2. The user can choose to search case sensitively and insensitively. 
3. The user can search up or down. 
4. Found text will be selected. 
5. The search will start from the active selected text (inclusive) if there is one, or where the cursor is. 
5. The search will automatically go around when it reaches the starting point. 

### Accessibility

> This feature could help multitaksing terminal users. 

Reviewer 2: Thankfully, we should benefit from the UIA provided directly from XAML. But we should consider how it's added to the UIA/XAML tree? Is the find box attached to the TermControl or to the TermApp?

As mentioned earlier, how can a user access all of this functionality without the use of a mouse? Are keybindings customizeable?

Our UIA providers need to be hooked up to this search functionality as well. That one isn't too difficult. Just a matter of putting what we need in the "FindText" function. I forget if it's in TermControlUiaProvider or UiaTextRange (I'd assume the first).

Reviewer: Super curious what this would look like from a screen reader perspective as well. When the text is found, what does Narrator say? @carlos-zamora  your thoughts and experience w/ conhost would be helpful here.

Author: My current design is that SearchBox is attached to TermControl.

Author: Selection is tested with Accessibility Insights, and it could correctly find the selection.

### Security

This feature should not introduce any new security issues.

> ### Reliability

Reviewer: Typically, the "reliability" section more refers to the tendency of the application to crash/hang/behave unexpectedly. Will this feature add any new cases we should be prepared for that could cause crashes?

Author: Added new concerns.

This feature enable users to search for text in the terminal input/output history. This is a widely-used feature in most editors and thus improve the reliability of Terminal once added. 

### Compatibility

This feature won't break existing features of Terminal.

### Performance, Power, and Efficiency

This feature only launches in need. It does not impact the performance of Terminal. 

> ## Potential Issues

Reviewer: What happens when the window isn't big enough for the full dialog to be visible?

Author: Oh, this is a issue that I should think about.

Author: The search box will be "squeezed":

![image](https://user-images.githubusercontent.com/11893272/69596072-3e2fb580-0fb6-11ea-9975-1442a4238c2d.png)

Author: Yes, it is similar to VSCode: 

![image](https://user-images.githubusercontent.com/11893272/69668381-a036fc00-1044-11ea-86b7-e62f5b546839.png)

Author: If the Terminal is too short, the Terminal screen will disappear though.

Author: Added

1. The search bar should not block command line view. 
2. Search should not block any program the Terminal is currently executing.

## Future considerations

In version 1, we want realize a case sensitive/insensitive exact text match. But we may consider the following features in version 2:

> 1. Add "Find" button in dropdown menu to trigger search. 

Reviewer 2: As an addition (or replacement) to this point, this may also be a good feature for the command palette #2046 . You and @zadjii-msft should talk about this one 😊

1. Search is performed on a XAML AutoSuggestBox. Once the user click the "Find" icon, we start from the cursor (the current input line on the Terminal) or the current selection, and try to find the exact text in the text buffer. The nearest searched one will be selected. And we set the search start point to the selected text. The next time "Find" button is clicked, the search will start before or after the previous searched text.
2. The user can choose to search up or down by choosing up arrow or down arrow buttons. A blue border will appear on the chosen button. 
3. The user can choose to do case sensitive or insensitive match by checking a check box. 
4. If the user click the "X" button, the search stopped and the search box disappears. In phrase one we do not store any state, but in the future we can consider storing the search history. 

## UI/UX Design

![Sol Design](images/SearchBoxControl.png)

Above is the SearchBoxControl. The two buttons with up/down arrows controls the search direction, a blue border will appear on the current selected direction. The checkbox, if checked, means that we are searching case-sensitivily. 

![Sol Design](images/SearchBoxUpSelected.png)

> The search box should be on the top right corner of the Terminal window. If the current tab is splitted into panes, each pane can have a individual searchbox. We need to avoid it blocking too much screen contents. 

Reviewer: ```suggestion
The search box should be on the top right corner of the Terminal window. If the current tab is split into panes, each pane can have a individual searchbox. We need to avoid it blocking too much screen contents. 
```

> The search box should be on the top right corner of the Terminal window. If the current tab is splitted into panes, each pane can have a individual searchbox. We need to avoid it blocking too much screen contents. 

Reviewer: > We need to avoid it blocking too much screen contents. 

How?

Our ultimate goal is to provide both search within one tab and search from all tabs experiences. But we can start with one-tab search. The search experience should have following features:

1. The search is triggered by KeyBindings. A new setting property named "openFind" will be enabled in the Json file. The user can set their own key bindings for search. The default is <kbd>ctrl+shift+f</kbd>. 
2. The user search in a XAML TextBox, which is contained in a custom `SearchBoxControl`. The default position of the search box is the top right corner, the user can also move it to bottom right corner to avoid blocking text. 
3. We can have multiple search methods. The simplest one is exact text match. Other match methods include case-sensitive exact match and regex match. In the first phrase, we will focus on case sensitive/insensitive text exact match. 
4. If currently there is no active selection, the search starts from the last line of the mutableViewport. If there is an active selection, we start from the previous or the next text of the selected text. We automatically go around if we reach the start point of the search. 

> 5. The search dialog should not block terminal's view. In phrase one, this is achieved by make search box movable. The search box could be moved to the bottom right by clicking a button. It will be moved back to the top right corner by clicking the button again. The button should be styled with the first click. 

Reviewer: > The button should be styled with the first click.

I don't think I know what this means - is the button highlighted when the button is clicked?

Author: Actually, for now the button will have the Font Icon upside-down.

3. We re-use the Search module in conhost. It performs the search in a brute-force approach. Starting from every position in the text buffer, the search algorithm compares the span of the searched string with buffer characters, and if the current buffer text matches the whole string, it will return store the position of the text in the buffer and return. The stored position information will be used for selection. 
3. The user can choose to search up or down. Search module realizes this, we just need to set a boolean flag. Default is search up. 
4. The user can choose to do case sensitive or insensitive match. This also realized by Search module by setting a boolean flag. Default is search case-insensitively. 
5. Tab navigation is realized by XAML. We just need to set TabNavigation="Cycle" in `SearchBoxControl`. 

> 6. If the user click the "X" button or press <kbd>Esc</kbd>, the search box will disappear and the object will be destructed and detached from the `TermControl` XAML tree. In phrase one we do not store any state. 

Reviewer: This line looks like it was moved down to L67, so it can probably be removed here

Author: I think this line focuses on the algorithmic perspective while line 67 focuses on the behavior that the user could be see.

We will create a SearchBoxControl Xaml UserControl element. When a search process begins, a SearchBoxControl object will be created and attach to TermControl root grid. In other words, one SearchBox is added for each TermControl. The reasons for this design is:

1. Each TermControl corresponds to an iterative Terminal Window and has a individual text buffer. In phrase 1 we are going to search witin the current terminal text buffer. 
2. If we put the search box under TerminalApp, then the search can only happen on the current focused Terminal. 
3. If the community does not like the current design, we can lift SearchBox to a higher level. 

Search process:
1. Search is performed on a XAML AutoSuggestBox. Once the user click the "Find" icon, we start from the cursor (the current input line on the Terminal) or the current selection, and try to find the exact text in the text buffer. The nearest searched one will be selected. And we set the search start point to the selected text. The next time "Find" button is clicked, the search will start before or after the previous searched text.

> 2. The user can choose to search up or down by choosing up arrow or down arrow buttons. A blue border will appear on the chosen button. 

Author: changed

> 3. The user can choose to do case sensitive or insensitive match by checking a check box. 

Reviewer: What's the default here?

Reviewer: (this doesn't look like it was actually answered...)

Author: Added default explanation.

4. If the user click the "X" button, the search stopped and the search box disappears. In phrase one we do not store any state, but in the future we can consider storing the search history. 

## UI/UX Design

> ![Sol Design](images/SearchBoxControl.png)

Reviewer: Make sure to change the "Sol Design" here to some text that's descriptive for each image

Author: changed

In version 1, we want realize a case sensitive/insensitive exact text match. But we may consider the following features in version 2:

> 1. Add "Find" button in dropdown menu to trigger search.Add "Find" button in dropdown menu to trigger search. This enables the search feature to be operated with mouse only. However, this is not required by Accessibility so we do not cover this in phase one. 

Reviewer 3: This text got duplicated...
